### PR TITLE
Fixes: #4597 Parse URI path with an endpoint

### PR DESCRIPTION
**Pull request by rynowak**
Adds functionality to LinkParser to parse a URI path given a way to
find an endpoint. This is the replacement for various machinications
using the global route collection and `RouteData.Routers` in earlier
versions.

For now I'm just adding a way to do this using Endpoint Name since it's
a pretty low level feature. Endpoint Name is also very direct, so it
feels good for something like this.

Example usage:

```C#
[HttpPut]
[Route("/Products/{id}", Name = "ProductsRoute")]
public async Task<IActionResult> AddRelatedProduct(int projectId, string path)
{
   var values = parser.ParsePathByEndpointName("ProductsRoute", path);
   var relatedId = (string)values["id"];

   // Do something with this information
}
```

### CODE REVIEW

**Comment by rynowak** [comment]
@JamesNK  looking for your COOL high level feedback on this before adding tests.

File: src/Http/Routing.Abstractions/src/LinkGenerator.cs [hunk 1]
```diff
@@ -147,5 +148,32 @@ public abstract class LinkGenerator
             PathString pathBase = default,
             FragmentString fragment = default,
             LinkOptions options = default);
+
+        /// <summary>
+        /// Attempts to pass the provided <paramref name="path"/> using the route pattern
```
Comment (rynowak) [inline src/Http/Routing.Abstractions/src/LinkGenerator.cs]: should say parse :)

File: src/Http/Routing/src/DefaultLinkGenerator.cs [hunk 2]
```diff
@@ -333,7 +388,7 @@ public static RouteValueDictionary GetAmbientValues(HttpContext httpContext)
 
         public void Dispose()
         {
-            _cache.Dispose();
+            _binderCache.Dispose();
```
Comment (rynowak) [inline src/Http/Routing/src/DefaultLinkGenerator.cs]: need to add my new cache here as well

**Comment by rynowak** [comment]
This is also pretty simple and straightforward compared to what we have today 👍

**Comment by rynowak** [comment]
There's a few more decisions here to be made like - does this apply route values? Does this run constraints? I'm planning to reach out to some of the folks that asked for this feature and poll them.

**Comment by rynowak** [comment]
@oferns @mgolois - I'm looking at adding an API to address your feedback here: https://github.com/aspnet/AspNetCore/issues/4597

Could you take a look at this and let me know what you think?

File: src/Http/Routing/src/DefaultLinkGenerator.cs [hunk 3]
```diff
@@ -280,13 +297,43 @@ private List<RouteEndpoint> GetEndpoints<TAddress>(TAddress address)
...
+            for (var i = 0; i < endpoints.Count; i++)
+            {
+                var endpoint = endpoints[i];
+                if (TryParse(endpoint, path, out var values))
+                {
+                    return values;
+                }
+            }
+
+            Log.LinkGenerationFailed(_logger, endpoints);
```
Comment (JamesNK) [inline src/Http/Routing/src/DefaultLinkGenerator.cs]: Wrong logging method

**Review by JamesNK — COMMENTED** [review]
Seems alright.

It does feel a little odd to have matching on link generator. Consider whether a new type makes sense. `LinkParser`?

File: src/Http/Routing/src/DefaultLinkGenerator.cs [hunk 4]
```diff
@@ -280,13 +297,43 @@ private List<RouteEndpoint> GetEndpoints<TAddress>(TAddress address)
             return null;
         }
 
+        public override RouteValueDictionary ParsePathByAddress<TAddress>(TAddress address, PathString path)
+        {
+            var endpoints = GetEndpoints(address);
+            if (endpoints.Count == 0)
```
Comment (oferns) [inline src/Http/Routing/src/DefaultLinkGenerator.cs]: How/where is GetEndpoints implemented? This was my issue, getting access to the routes/endpoints that used to be in RouteData.Routes

Comment (rynowak) [reply src/Http/Routing/src/DefaultLinkGenerator.cs]: Get endpoints is implemented on the base class. 

If you want a list of all of the endpoints in the app, you do do that by grabbing `EndpointDataSource` from DI. If you want the endpoint that matched the current request, use this.  https://github.com/aspnet/AspNetCore/blob/425c196cba530b161b120a57af8f1dd513b96f67/src/Http/Http.Abstractions/src/Routing/EndpointHttpContextExtensions.cs

**Comment by rynowak** [comment]
@JamesNK - ready for review. I decided to go with a separate class.

File: src/Http/Routing/src/DefaultLinkParser.cs [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+            }
+            else
+            {
+                Log.EndpointsFound(_logger, address, endpoints);
+            }
+
+            return endpoints;
+        }
+
+
```
Comment (JamesNK) [inline src/Http/Routing/src/DefaultLinkParser.cs]: too liney

File: src/Http/Routing/src/DefaultLinkParser.cs [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+            public static class EventIds
+            {
+                public static readonly EventId EndpointsFound = new EventId(100, "EndpointsFound");
+                public static readonly EventId EndpointsNotFound = new EventId(101, "EndpointsNotFound");
+
+                public static readonly EventId PathParsingSucceeded = new EventId(102, "PathParsingSucceeded");
+                public static readonly EventId PathParsingFailed = new EventId(103, "PathParsingFailed");
+            }
+
+            private static readonly Action<ILogger, IEnumerable<string>, object, Exception> _endpointsFound = LoggerMessage.Define<IEnumerable<string>, object>(
```
Comment (JamesNK) [inline src/Http/Routing/src/DefaultLinkParser.cs]: I didn't know logging supported enumerable. Could you add a unit test that checks that the log message looks like what you expected when there are multiple matches.

Alternatively I think it is fine to just have a count in the log message.

File: src/Http/Routing/test/UnitTests/DefaultLinkParserTest.cs [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+
+            // Act
+            var values = parser.ParsePathByAddress(0, "/");
+
+            // Assert
+            MatcherAssert.AssertRouteValuesEqual(new { controller = "Home", action = "Index", }, values);
+        }
+
+        [Fact]
+        public void ParsePathByAddresss_HasMatches_RunsConstraints()
```
Comment (JamesNK) [inline src/Http/Routing/test/UnitTests/DefaultLinkParserTest.cs]: 👍

File: src/Http/Routing/src/LinkParserEndpointNameAddressExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+    /// <summary>
+    /// Extension methods for using <see cref="LinkParser"/> with an endpoint name.
+    /// </summary>
+    public static class LinkParserEndpointNameAddressExtensions
+    {
+        /// <summary>
+        /// Attempts to parse the provided <paramref name="path"/> using the route pattern
+        /// specified by the <see cref="Endpoint"/> matching <paramref name="endpointName"/>.
+        /// </summary>
+        /// <param name="parser">The <see cref="LinkGenerator"/>.</param>
```
Comment (JamesNK) [inline src/Http/Routing/src/LinkParserEndpointNameAddressExtensions.cs]:
```suggestion
        /// <param name="parser">The <see cref="LinkParser"/>.</param>
```

File: src/Http/Routing/src/DefaultLinkParser.cs [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+            {
+                return null;
+            }
+
+            for (var i = 0; i < endpoints.Count; i++)
+            {
+                var endpoint = endpoints[i];
+                if (TryParse(endpoint, path, out var values))
+                {
+                    Log.PathParsingSucceeded(_logger, path, endpoints);
```
Comment (JamesNK) [inline src/Http/Routing/src/DefaultLinkParser.cs]: Should this log message just mention the endpoint that was successful?

File: src/Http/Routing/src/DefaultLinkParser.cs (hunk 6 above)
Comment (rynowak) [reply src/Http/Routing/src/DefaultLinkParser.cs]: I did know that  😆

File: src/Http/Routing/src/LinkParserEndpointNameAddressExtensions.cs (hunk 8 above)
Comment (rynowak) [reply src/Http/Routing/src/LinkParserEndpointNameAddressExtensions.cs]: Ooops!

**Comment by rynowak** [comment]
@JamesNK updated

File: src/Http/Routing/src/DefaultLinkParser.cs [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNetCore.Routing
+{
+    internal class DefaultLinkParser : LinkParser, IDisposable
```
Comment (JamesNK) [inline src/Http/Routing/src/DefaultLinkParser.cs]: tis never registered with DI

Add an integration test that uses it

**Comment by rynowak** [comment]
@JamesNK updated

File: src/Mvc/test/Mvc.FunctionalTests/LinkParserTest.cs [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+                p =>
+                {
+                    Assert.Equal("action", p.Name);
+                    Assert.Equal("AnotherRoute", p.Value.Value<string>());
+                },
+                p =>
+                {
+                    Assert.Equal("controller", p.Name);
+                    Assert.Equal("LinkParser", p.Value.Value<string>());
+                },
```
Comment (rynowak) [inline src/Mvc/test/Mvc.FunctionalTests/LinkParserTest.cs]: The fact that controller/action are included here (non-parameter defaults) might seem surprising at first glance but I think the alternative is WAYYYY more surprising. 

You really want this to round-trip the values from link generation and omitting non-parameter defaults would be strange and arbitrary

**Review by JamesNK — APPROVED** [review]

**Comment by marono** [comment]
Will this make it into .NET Core 3.0? Thanks!

**Comment by rynowak** [comment]
This is already in 3.0 preview 6 which will be the next preview release.

**Comment by holytshirt** [comment]
Having been looking how to do this in 2.2 and just found this :/. 

Hurry up 3.0 and thank you!
This is a big missing piece when doing a proper ReST apis, as you don't post Ids, you post uris.

**Comment by rynowak** [comment]
@holytshirt - thanks.

I'm surprised we took this long to figure this out! I've been doing work on routing now for a few years, and I saw a big outcry from users when I broke the existing way of doing this 😆 - now this is something proper that is in the docs.

**Comment by gbrunton** [comment]
@rynowak  Where in the docs is this because I'm having a hard time find it?

**Comment by rynowak** [comment]
It's mentioned here: https://docs.microsoft.com/en-us/aspnet/core/migration/22-to-30?view=aspnetcore-3.1&tabs=visual-studio#routing-startup-code

I've opened #16913 to track adding some more information.

**Pull request merged**